src/op_analytics/cli/subcommands/pulls/agora/fetch.py

```diff
@@ -0,0 +1,116 @@
+from typing import List
+import os
+from op_analytics.coreutils.logger import structlog
+from op_analytics.coreutils.request import get_data
+
+
+log = structlog.get_logger()
+
+BASE_URL = "https://vote.optimism.io/api/v1"
+ENDPOINTS = {
+    "delegates": f"{BASE_URL}/delegates",
+    "proposals": f"{BASE_URL}/proposals",
+    "proposal_by_id": f"{BASE_URL}/proposals/{{proposal_id}}",
+    "votes_by_proposal": f"{BASE_URL}/proposals/{{proposal_id}}/votes",
+    "delegatees": f"{BASE_URL}/delegates/{{addressOrEnsName}}/delegatees",
+}
+
+AUTH_TOKEN = os.getenv("AGORA_API_KEY")
+
+if not AUTH_TOKEN:
+    raise EnvironmentError("AGORA_API_KEY environment variable is not set.")
+
+HEADERS = {
+    "Authorization": f"Bearer {AUTH_TOKEN}",
+    "Content-Type": "application/json",
+}
+
+log = structlog.get_logger()
+
+
+# Todo: Transport to coreutils
+def fetch_paginated_data_with_offset(
+    session, url: str, offset: int = 0, params: dict = {}
+) -> List[dict]:
+    data = []
+    limit = params.get("limit", 100)
+    offset = offset
+
+    while True:
+        query_params = {"limit": limit, "offset": offset}
+        if params:
+            query_params.update(params)
+
+        response = get_data(session, url, headers=HEADERS, params=query_params)
+
+        # Check if the response is a list or a dictionary
+        if isinstance(response, list):
+            batch = response
+        else:
+            batch = response.get("data", [])
+
+        data.extend(batch)
+
+        # If the response is a list, we assume there's no pagination metadata
+        if isinstance(response, list) or not response.get("metadata", {}).get("has_next", False):
+            break
+
+        # Update offset for the next batch
+        offset = response.get("metadata", {}).get("next_offset", offset + limit)
```

Comment:
I'm curious about the fallbacks. If we don't have metadata, should we fail instead? I'm assuming that not having "metadata" would be the API behaving in an unexpected way, so working around it could lead to data integrity problems for the ingested data.